File: rust/src/code/remove_all_adjacent_duplicates_in_string_ii.rs
struct Solution {}

impl Solution {
    // NOTE: tmie complexity: O(n)
    pub fn remove_duplicates(s: String, k: i32) -> String {
        let mut stack: Vec<(char, i32)> = Vec::new();
        let s: Vec<char> = s.chars().collect();
        for i in 0..s.len() {
            let mut count: i32 = 1;
            if stack.len() > 0 && stack[stack.len() - 1].0 == s[i] {
                count += stack[stack.len() - 1].1;
            }
            if count == k {
                while stack.len() > 0 && stack[stack.len() - 1].0 == s[i] {
                    stack.pop().unwrap();
                    count -= 1;
                }
            } else {
                stack.push((s[i], count));
            }
        }

        stack.iter().map(|x| x.0).collect::<String>()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_remove_duplicates_case_1() {
        // arrange
        let s: String = "abcd".to_string();
        let k: i32 = 2;
        let expected: String = "abcd".to_string();

        // act
        let actual = Solution::remove_duplicates(s, k);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_remove_duplicates_case_2() {
        // arrange
        let s: String = "deeedbbcccbdaa".to_string();
        let k: i32 = 3;
        let expected: String = "aa".to_string();

        // act
        let actual = Solution::remove_duplicates(s, k);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_remove_duplicates_case_3() {
        // arrange
        let s: String = "pbbcggttciiippooaais".to_string();
        let k: i32 = 2;
        let expected: String = "ps".to_string();

        // act
        let actual = Solution::remove_duplicates(s, k);

        // assert
        assert_eq!(expected, actual);
    }
}

File: python/remove_all_adjacent_duplicates_in_string_ii.py
class Solution:
    def removeDuplicates(self, s: str, k: int) -> str:
        """time complexity: O(n), space complexity: O(n)"""
        stack: list[tuple[str, int]] = list()
        for i in range(len(s)):
            count: int = 1
            if len(stack) > 0 and stack[-1][0] == s[i]:
                count += stack[-1][1]
            if count == k:
                while len(stack) > 0 and stack[-1][0] == s[i]:
                    stack.pop()
            else:
                stack.append((s[i], count))
        return "".join([s[0] for s in stack])


def test_removeDuplicates_case_1():
    # arrange
    s: str = "abcd"
    k: int = 2
    expected: str = "abcd"

    # act
    solution = Solution()
    actual = solution.removeDuplicates(s, k)

    # assert
    assert expected == actual


def test_removeDuplicates_case_2():
    # arrange
    s: str = "deeedbbcccbdaa"
    k: int = 3
    expected: str = "aa"

    # act
    solution = Solution()
    actual = solution.removeDuplicates(s, k)

    # assert
    assert expected == actual


def test_removeDuplicates_case_3():
    # arrange
    s: str = "pbbcggttciiippooaais"
    k: int = 2
    expected: str = "ps"

    # act
    solution = Solution()
    actual = solution.removeDuplicates(s, k)

    # assert
    assert expected == actual


def test_removeDuplicates_case_4():
    # arrange
    s: str = "yfttttfbbbbnnnnffbgffffgbbbbgssssgthyyyy"
    k: int = 4
    expected: str = "ybth"

    # act
    solution = Solution()
    actual = solution.removeDuplicates(s, k)

    # assert
    assert expected == actual
